util/topsql/tracecpu/profile.go

```diff
@@ -0,0 +1,421 @@
+// optimizer takes time to generated plan.
+// After this tune function, the `sqlStats` become to:
+//     plans: {
+//         "optimize"  : 100ms, // 600 - 200 - 300
+//         "table_scan": 200ms,
+//         "index_scan": 300ms,
+//       },
+//     total:      600ms,
+func (s *sqlStats) tune() {
+	if len(s.plans) == 0 {
+		s.plans[""] = s.total
+		return
+	}
+	planTotal := int64(0)
+	for _, v := range s.plans {
+		planTotal += v
+	}
+	optimize := s.total - planTotal
+	if optimize <= 0 {
+		return
+	}
+	s.plans[""] += optimize
+}
+
+func (sp *topSQLCPUProfiler) handleExportProfileTask(p *profile.Profile) {
+	sp.mu.Lock()
+	defer sp.mu.Unlock()
+	if sp.mu.ept == nil {
+		return
+	}
+	sp.mu.ept.mergeProfile(p)
+}
+
+func (sp *topSQLCPUProfiler) hasExportProfileTask() bool {
+	sp.mu.Lock()
+	has := sp.mu.ept != nil
+	sp.mu.Unlock()
+	return has
+}
+
+// IsEnabled return true if it is(should be) enabled. It exports for tests.
+func (sp *topSQLCPUProfiler) IsEnabled() bool {
+	return variable.TopSQLEnabled() || sp.hasExportProfileTask()
+}
+
+// StartCPUProfile same like pprof.StartCPUProfile.
+// Because the GlobalTopSQLCPUProfiler keep calling pprof.StartCPUProfile to fetch SQL cpu stats, other place (such pprof profile HTTP API handler) call pprof.StartCPUProfile will be failed,
+// other place should call tracecpu.StartCPUProfile instead of pprof.StartCPUProfile.
+func StartCPUProfile(w io.Writer) error {
+	if GlobalTopSQLCPUProfiler.IsEnabled() {
+		return GlobalTopSQLCPUProfiler.startExportCPUProfile(w)
+	}
+	return pprof.StartCPUProfile(w)
+}
+
+// StopCPUProfile same like pprof.StopCPUProfile.
+// other place should call tracecpu.StopCPUProfile instead of pprof.StopCPUProfile.
+func StopCPUProfile() error {
+	if GlobalTopSQLCPUProfiler.IsEnabled() {
+		return GlobalTopSQLCPUProfiler.stopExportCPUProfile()
+	}
+	pprof.StopCPUProfile()
+	return nil
+}
+
+// ResetGoroutineLabelsWithOriginalCtx resets the goroutine label with the original ctx.
+func ResetGoroutineLabelsWithOriginalCtx(ctx context.Context) {
+	pprof.SetGoroutineLabels(ctx)
+}
+
+// SetSQLLabels sets the SQL digest label into the goroutine.
+func SetSQLLabels(ctx context.Context, normalizedSQL, sqlDigest string) context.Context {
+	if variable.EnablePProfSQLCPU.Load() {
+		ctx = pprof.WithLabels(ctx, pprof.Labels(labelSQLDigest, sqlDigest, labelSQL, util.QueryStrForLog(normalizedSQL)))
+	} else {
+		ctx = pprof.WithLabels(ctx, pprof.Labels(labelSQLDigest, sqlDigest))
+	}
+	pprof.SetGoroutineLabels(ctx)
+	return ctx
+}
+
+// SetSQLAndPlanLabels sets the SQL and plan digest label into the goroutine.
+func SetSQLAndPlanLabels(ctx context.Context, sqlDigest, planDigest string) context.Context {
+	ctx = pprof.WithLabels(ctx, pprof.Labels(labelSQLDigest, sqlDigest, labelPlanDigest, planDigest))
+	pprof.SetGoroutineLabels(ctx)
+	return ctx
+}
+
+func (sp *topSQLCPUProfiler) startExportCPUProfile(w io.Writer) error {
+	sp.mu.Lock()
```

Comment:
This can be replaced by atomic.Value?

Reply:
No, the mutex also use to protect those function and other function.